### PR TITLE
Terminal ansi colors

### PR DESCRIPTION
**Pull request by DR-Reg**
Adds support for using ansi escape codes to represent different severity levels when printing to a terminal, and presents some default escape codes. Allows for . Solves #103.

Unit tests were not written however, as I was unsure how to capture the stdout to check it was correctly printing ansi terminals (I did check with a script on my side that it was working).

This is my first open source contribution so any feedback is welcome.

### CODE REVIEW

**Comment by DR-Reg** [comment]
Do I need to 'sign the commit' so it can be merged? Or is anything else missing?

**Comment by goatshriek** [comment]
Sorry for the delay in response, I have been focused on a different project for a bit.

First of all, thank you for tackling this issue! As I am sure you saw in the issue history it has drawn many contenders but remains yet unconquered.

Signing commits is not a requirement for PRs here, particularly since they are typically squashed anyway. However, I do recommend you work through that process for your own education (if education is something you're into, at least). But don't worry about signing and re-pushing your work or anything like that.

At first glance this seems like a solid contribution that shouldn't need much re-shaping, but I need to set aside time to sit down and review it in detail. I will get back to you within the next week with a review.

File: include/stumpless/target/stream.h
Comment (goatshriek) [inline include/stumpless/target/stream.h]: update the race variable name to `escape_code`

File: include/stumpless/target/stream.h
Comment (goatshriek) [inline include/stumpless/target/stream.h:158]: This function documentation is a good place to include a link to information on ANSI escape codes, perhaps the Wikipedia article for "ANSI escape code"?

File: src/target/stream.c
Comment (goatshriek) [inline src/target/stream.c]: use `severity_is_invalid` here instead

File: src/target/stream.c
Comment (goatshriek) [inline src/target/stream.c]: similarly, use `raise_invalid_severity` here instead

File: src/target/stream.c
Comment (goatshriek) [inline src/target/stream.c]: acquire this lock before checking the type, so that it is locked for the duration of the code that uses it

File: include/stumpless/target/stream.h
Comment (goatshriek) [inline include/stumpless/target/stream.h]: this is both `AS-Unsafe` and `AC-Unsafe` due to lock usage, but I don't see any heap usage. Be sure you update the verbiage below the designations as well.

File: src/target/stream.c
Comment (goatshriek) [inline src/target/stream.c]: you'll probably need to expand this from three individual critical sections to a single larger one, locking before the first write and only unlocking after the reset code is written. Otherwise, writes may end up being interleaved between messages with strange resulting output.

File: src/target/stream.c
Comment (goatshriek) [inline src/target/stream.c]: should the argument to `strlen` be `reset_code` here instead?

File: src/target/stream.c
Comment (goatshriek) [inline src/target/stream.c:143]: be sure you call `clear_error` before returning so that the error code is set to clean in the event of success

File: src/target/stream.c
Comment (goatshriek) [inline src/target/stream.c:132]: This error message should be a localized string. The documentation for [localization](https://github.com/goatshriek/stumpless/blob/latest/docs/localization.md#adding-new-strings) describes how to add new localized strings: you just need to run `scripts/add_l10n_string.rb` with your new string.

**Review by goatshriek — CHANGES_REQUESTED** [review]
Thanks again for putting this together, and your patience in waiting for a review. This is great work! The changes below should be relatively straightforward.

However, you will also need to write some tests for this functionality. This will be a little more involved - the way I would approach this is to create a file, and use the stream to create a new stream target that writes into it. You can then check the file contents after the writing to make sure that the expected escape codes are there (or not, depending on the test).

File: include/stumpless/target/stream.h
Comment (DR-Reg) [reply include/stumpless/target/stream.h]: Done.

File: include/stumpless/target/stream.h
Comment (DR-Reg) [reply include/stumpless/target/stream.h:158]: Done.

File: src/target/stream.c
Comment (DR-Reg) [reply src/target/stream.c]: Done.

File: src/target/stream.c
Comment (DR-Reg) [reply src/target/stream.c]: Done.

File: src/target/stream.c
Comment (DR-Reg) [reply src/target/stream.c]: Done.

File: include/stumpless/target/stream.h
Comment (DR-Reg) [reply include/stumpless/target/stream.h]: Done.

File: src/target/stream.c
Comment (DR-Reg) [reply src/target/stream.c]: Done.

File: src/target/stream.c
Comment (DR-Reg) [reply src/target/stream.c]: Yes. Done.

File: src/target/stream.c
Comment (DR-Reg) [reply src/target/stream.c:143]: Done.

File: src/target/stream.c
Comment (DR-Reg) [reply src/target/stream.c:132]: Thank you for all the feedback! I have been unable to add the localised string due to what may be a bug in the script (output shown below). I am not very fluent in ruby so I decided to just leave it as is for somebody else to work on.

Output:
```
$ ruby scripts/add_l10n_string.rb SET_SEVERITY_COLOR_UNSUPPORTED_TARGET_ERROR_MESSAGE This function is only supported for stream targets
scripts/add_l10n_string.rb:59:in `each': Is a directory @ io_fillbuf - fd:7 <home-path>/stumpless/include/private/config/locale/. (Errno::EISDIR)
        from scripts/add_l10n_string.rb:59:in `block in <main>'
        from scripts/add_l10n_string.rb:53:in `each'
        from scripts/add_l10n_string.rb:53:in `<main>'
```

Edit: forgot to mention it is ruby 3.3

Comment (goatshriek) [reply src/target/stream.c:132]: Ah, that is unfortunate. Thank you very much for sharing the details though, they did allow me to identify the issue and a fix. You can leave this out and I will go back and correct it afterwards - line 53 of the script needs to be changed to `each_child`. I'll submit a separate patch later.

**Comment by goatshriek** [comment]
You're closing in on done! Just two more changes needed - you'll need to resolve the Windows test failures, which will probably require some refactoring of the tests themselves to be a bit more portable. You may need to create a new file instead of using the stderr/stdout targets, but I'll leave this up to you.

Also, please add a test where you attempt to call `stumpless_set_severity_color` with an incompatible target type, to make sure that this fails as expected.

**Comment by DR-Reg** [comment]
I have added some preprocessor directives so the syscalls to dup etc. are conformant with the required windows ISO C rules. In the end I thought it was better to redirect stdout/stderr since in theory (as I understand) files should not have these escape codes (unless set manually).

I am aware that it may be possible to make this code simpler using some sort of macro to distinguish between code needed for windows vs unix. (however I wasn't sure what your policy on macros is, I can rewrite the code in this way if you want).

Also if you think the file approach is cleaner I will rewrite the test.

Also note that for some reason running on my windows machine I get that the regex expression is invalid? Is this a bug in the windows version of gtest or my fault?

Sorry for the inconvenience.

Edit: PS. Should the test for `stumpless_set_severity_color` be in test/function/target/stream.cpp or elsewhere?

**Comment by goatshriek** [comment]
The tests have a more relaxed standard for macros and wrapping up OS-specific functionality, so I think what you've done here is fine, no need to rewrite at this point. As long as they run on the major tested systems (which are currently the ones covered by CI) then I think we're good to go.

Hmm, I'm not sure why the Windows regexes are not working properly - perhaps the underlying implementation is more strict about regular expressions with non-displayable characters in them? Does changing the escape character to `\x1b` instead of the octal value help?

Yes, I think that `test/function/target/stream.cpp` is the best place for the set severity color tests.

**Comment by DR-Reg** [comment]
In the end the regex bug was due to gtest having limited regex support on windows (as seen [here](https://google.github.io/googletest/advanced.html#regular-expression-syntax)). I assume they will add better support in the future, so I put in a bit of a 'janky' solution for windows using one of GTEST's processor macros (so when they do add the functionality the actual regex is called).
Is this ok?

**Comment by goatshriek** [comment]
This will be okay, though I will ask that you please add a comment to the test code before the preprocessor conditional that explains why this is done so that others reading the tests will know why this is being done without finding this conversation.

It looks like you just need to update the `src/windows/stumpless.def` file with the new function, and address some memory leaks that valgrind has identified, and you'll be closing in on done.

**Review by goatshriek — APPROVED** [review]

**Comment by goatshriek** [comment]
Thank you again for working through this, and I hope that you feel a sense of accomplishment for completing an issue that gave so many others quite a bit of trouble!

I try to give contributors a shoutout on X when I can, especially for larger contributions like this one. I don't see one listed on your profile here though: if you have an account let me know (via email or private Gitter if you want it to remain hidden) and I will post something there.

**Pull request merged**
